spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkExecutorCache.java

```diff
@@ -0,0 +1,197 @@
+import org.apache.iceberg.relocated.com.google.common.collect.Queues;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An executor cache for optimizing tasks by reducing the computation and IO overhead.
+ *
+ * <p>The cache is configurable and enabled through Spark SQL properties. Its key features include
+ * setting limits on the total cache size and maximum size for individual entries. Additionally, it
+ * implements automatic eviction of entries after a specified duration of inactivity. The cache will
+ * respect the SQL configuration valid at the time of initialization. All subsequent changes will
+ * have no effect.
+ *
+ * <p>Usage pattern involves fetching data from the cache using a unique combination of execution ID
+ * and key. If the data is not present in the cache, it is computed using the provided supplier and
+ * stored in the cache, subject to the defined size constraints.
+ *
+ * <p>Note that this class employs the singleton pattern to ensure only one cache exists per JVM.
+ *
+ * @see SparkUtil#executionId()
+ */
+public class SparkExecutorCache {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SparkExecutorCache.class);
+
+  private static final SparkConfParser CONF_PARSER = new SparkConfParser();
+  private static final boolean CACHE_ENABLED = parseCacheEnabledConf();
+  private static final Duration TIMEOUT = parseTimeoutConf();
+  private static final long MAX_ENTRY_SIZE = parseMaxEntrySizeConf();
+  private static final long MAX_TOTAL_SIZE = parseMaxTotalSizeConf();
+  private static final String EXECUTOR_DESC = SparkUtil.executorDesc();
+
+  private static volatile SparkExecutorCache instance = null;
+
+  private final Map<String, Collection<String>> keysByExecutionId;
+  private volatile Cache<String, CacheValue> cache = null;
+
+  private SparkExecutorCache() {
+    this.keysByExecutionId = Collections.synchronizedMap(Maps.newHashMap());
+  }
+
+  public static SparkExecutorCache getOrCreate() {
+    if (instance == null && CACHE_ENABLED) {
+      synchronized (SparkExecutorCache.class) {
+        if (instance == null) {
+          SparkExecutorCache.instance = new SparkExecutorCache();
+        }
+      }
+    }
+
+    return instance;
+  }
+
+  public static void cleanUp(String executionId) {
+    if (instance != null) {
+      instance.invalidate(executionId);
+    }
+  }
+
+  public long maxEntrySize() {
+    return MAX_ENTRY_SIZE;
+  }
+
+  public <V> V get(String executionId, String key, Supplier<V> valueSupplier, long valueSize) {
+    if (valueSize > MAX_ENTRY_SIZE) {
+      return valueSupplier.get();
+    }
+
+    storeMapping(executionId, key);
+
+    CacheValue value = cache().get(key, ignored -> new CacheValue(valueSupplier, valueSize));
+    Preconditions.checkNotNull(value, "Loaded value must not be null");
+    return value.get();
+  }
+
+  private void storeMapping(String executionId, String key) {
+    Collection<String> keys =
```

Comment:
optional:  is there any need to define variable keys, can we just inline the whole statement:

keysByExecutionId
   .computeIfAbsent(executionId, id -> Queues.newConcurrentLinkedQueue())
   .add(key)

Reply:
Applied.